tfsdk/tftypes_value.go

```diff
@@ -0,0 +1,221 @@
+package tfsdk
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+// createParentValue ensures that the given parent value can have children
+// values upserted. If the parent value is known and not null, it is returned
+// without modification. A null Object or Tuple is converted to known with null
+// children. An unknown Object or Tuple is converted to known with unknown
+// children. List, Map, and Set are created with empty elements.
+//
+// The parentType parameter ensures that the value will match the current
+// schema, not what is currently stored in the parentValue.
+func createParentValue(ctx context.Context, parentPath *tftypes.AttributePath, parentType tftypes.Type, parentValue tftypes.Value) (tftypes.Value, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	if !parentValue.IsNull() && parentValue.IsKnown() {
+		return parentValue, diags
+	}
+
+	var childValue interface{}
+
+	if !parentValue.IsKnown() {
+		childValue = tftypes.UnknownValue
+	}
+
+	switch parentType := parentType.(type) {
+	case tftypes.List:
+		parentValue = tftypes.NewValue(parentType, []tftypes.Value{})
+	case tftypes.Set:
+		parentValue = tftypes.NewValue(parentType, []tftypes.Value{})
+	case tftypes.Map:
+		parentValue = tftypes.NewValue(parentType, map[string]tftypes.Value{})
+	case tftypes.Object:
+		vals := map[string]tftypes.Value{}
+
+		for name, t := range parentType.AttributeTypes {
+			vals[name] = tftypes.NewValue(t, childValue)
+		}
+
+		parentValue = tftypes.NewValue(parentType, vals)
+	case tftypes.Tuple:
+		vals := []tftypes.Value{}
+
+		for _, elementType := range parentType.ElementTypes {
+			vals = append(vals, tftypes.NewValue(elementType, childValue))
+		}
+
+		parentValue = tftypes.NewValue(parentType, vals)
+	default:
+		diags.AddAttributeError(
+			parentPath,
+			"Value Conversion Error",
+			"An unexpected error was encountered trying to create a value. This is always an error in the provider. Please report the following to the provider developer:\n\n"+
+				fmt.Sprintf("Unknown parent type %T to create value.", parentValue.Type()),
```

Comment:
```suggestion
				fmt.Sprintf("Unknown parent type %s to create value.", parentValue.Type()),
```

This will give us a friendlier output of what the type actually is semantically, not what the Go implementation of it is.